polkadot/xcm/xcm-builder/src/routing.rs

```diff
@@ -231,3 +232,53 @@ impl<Inner: SendXcm> SendXcm for EnsureDecodableXcm<Inner> {
 		Inner::ensure_successful_delivery(location);
 	}
 }
+
+/// An adapter implementation of `SendXcm` for message queues that handles sending XCM messages.
+///
+/// Validates a message before sending by:
+/// 1. Converting the destination to an origin
+/// 2. Encoding the XCM message within size limits
+///
+/// Delivers by enqueuing the message in the queue.
+///
+/// Generic parameters:
+/// - `Queue`: The message queue implementation that can enqueue messages
+/// - `Origin`: The origin type used for message sources/destinations
+/// - `Conversion`: Type that can convert between Location and Origin
+pub struct MessageQueueRouterFor<Queue, Origin, Conversion>(
+	PhantomData<(Queue, Origin, Conversion)>,
+);
+impl<
+		Queue: EnqueueMessage<Origin>,
+		Origin: MaxEncodedLen,
+		Conversion: for<'a> MaybeConvert<&'a Location, Origin>,
+	> SendXcm for MessageQueueRouterFor<Queue, Origin, Conversion>
+{
+	/// The ticket type containing the encoded ('VersionedXcm') message and origin.
+	type Ticket = (BoundedVec<u8, Queue::MaxMessageLen>, Origin);
+
+	fn validate(
+		dest: &mut Option<Location>,
+		msg: &mut Option<Xcm<()>>,
+	) -> SendResult<Self::Ticket> {
+		// Convert dest to origin.
+		let origin = match dest.as_ref().and_then(Conversion::maybe_convert) {
+			Some(origin) => origin,
+			_ => return Err(SendError::NotApplicable),
+		};
+
+		// We can consume a message now.
+		let xcm = VersionedXcm::from(msg.take().ok_or(SendError::MissingArgument)?);
+		// We need to respect `Queue::MaxMessageLen`.
+		let bounded =
+			BoundedVec::try_from(xcm.encode()).map_err(|_| SendError::ExceedsMaxMessageSize)?;
+
+		Ok(((bounded, origin), Assets::new()))
+	}
+
+	fn deliver((msg, origin): Self::Ticket) -> Result<XcmHash, SendError> {
+		let unique_id = unique((&msg, &origin));
+		Queue::enqueue_message(msg.as_bounded_slice(), origin);
+		Ok(unique_id)
```

Comment:
@raymondkfcheung probably like this?

```suggestion
		let hash = sp_io::hashing::blake2_256(&msg);
		Queue::enqueue_message(msg.as_bounded_slice(), origin);
		Ok(hash)
```